### PR TITLE
Story/RETAILER-3370/ New auth services

### PR DESCRIPTION
**Pull request by mkmbayona**
#### Description

https://ordermentum.atlassian.net/browse/RETAILER-3370

Add new auth services
Remove unused models

----
#### PR Type

_Check one or more and add the corresponding number of reviewers_

- [ ] Bugfix _(1)_
- [ ] Minor change _(1)_
- [ ] Refactor (no functional changes/no api changes) _(1)_
- [ ] Feature _(2)_
- [ ] Backwards compatible change in functionality _(2)_
- [ ] Breaking change (existing functionality no longer works as expected) _(2)_
- [ ] Critical impact (security, payments or critical functionality) _(2+CTO)_

----
#### Changes

_List the main changes in this PR. Include screenshots if necessary._

----

#### Acceptance Criteria

_How can we test that this meets the requirements of the ticket?_

----

#### Checklist:

- [ ] Tests – Have you added or edited the test cases?
- [ ] Context – Have you tested this with a non-admin user?
- [ ] Rebase – Has this PR been rebased against `develop`?
- [ ] Feature Flag - If this is a new feature, does it have a feature flag?
- [ ] Release - Can this be released to production once tested?

### CODE REVIEW

File: Sources/OrdermentumSDK/Models/Auth.swift [hunk 1]
```diff
@@ -59,3 +59,91 @@ public struct ResendVerifyEmailRequest: Codable {
...
+
+public struct AuthAccess: Decodable {
+    public var id: String = ""
+    public var token: String = ""
+    public var accessToken: String = ""
+    public var refreshToken: String = ""
+    public var expiresIn: Double = 0
+    public var tokenType: String = ""
+    
+    public init() {}
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Models/Auth.swift:138]: Can you move the init above the var definitions

File: Sources/OrdermentumSDK/Models/Cart.swift [hunk 2]
```diff
@@ -8,15 +8,6 @@
...
 
-public struct Cart {
-    public init(deliveryDate: String?, lineItems: [CartLineItem]?, retailerId: String?, supplierId: String?) { }
-
-    public var deliveryDate: String?
-    public var lineItems: [CartLineItem]?
-    public var retailerId: String?
-    public var supplierId: String?
-}
-
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Models/Cart.swift:19]: Why does this PR touch the cart? Not sure we want to be deleting the cart here

File: Sources/OrdermentumSDK/Models/Cart.swift [hunk 3]
```diff
@@ -33,21 +24,6 @@ public struct CartProduct {
...
-    public var retailerId: String?
-    public var deliveryDate: String?
-    public var comment: String?
-    public var reference: String?
-    public var status: String?
-    public var type: String?
-    public var paymentMethodId: String?
-    public var origin: String?
-}
-
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Models/Cart.swift:50]: Again, not sure why we're deleting this? Are they unused?

File: Sources/OrdermentumSDK/Models/CodingKeys.swift [hunk 4]
```diff
@@ -621,4 +621,17 @@ enum CodingKeys: String, CodingKey {
...
+    case client_id
+    case client_secret
+    case grant_type
+    case username
+    case password
+    case refresh_token
+    // MARK: Auth response
+    case access_token
+    case expires_in
+    case token_type
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Models/CodingKeys.swift:636]: For these cases specifically, fine to leave as camel case as that's what the server epxects.

File: Sources/OrdermentumSDK/Routers/AuthRouter.swift [hunk 5]
```diff
@@ -95,13 +94,22 @@ public enum AuthRouter: URLRequestConvertible {
...
     }
     
     //Builder
     public func asURLRequest() throws -> URLRequest {
-        return try Client.instance.urlRequest(path: version+path, method: method, parameters: parameters, body: body)
+        switch self {
+        case .requestToken, .revokeToken:
+            return try AuthClient.instance.urlRequest(path: version+path, method: method, parameters: parameters, body: body)
+        default:
+            return try Client.instance.urlRequest(path: version+path, method: method, parameters: parameters, body: body)
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Routers/AuthRouter.swift:112]: Can avoid this by using `OM`. See `DocumentumService`

File: Sources/OrdermentumSDK/Services/AuthService.swift [hunk 6]
```diff
@@ -108,4 +108,55 @@ public class AuthService {
             completion(result, errorObject)
         }
     }
+    
+    
+}
+
+public class AuthServiceV2 {
+    public typealias AuthResponseCompletion = (Bool, AuthAccess?, ErrorResponse?) -> Void
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Services/AuthService.swift:116]: Can we please remove this typealias. I understand the cleaniliness however it implies that all calls that relate to auth will always return this.

File: Sources/OrdermentumSDK/Utilities/Config.swift [hunk 7]
```diff
@@ -0,0 +1,35 @@
+//
+//  File.swift
+//  
+//
+//  Created by Mark Kenneth Bayona on 9/2/21.
+//
+
+import Foundation
+
+struct K {
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Utilities/Config.swift:10]: What is `K`? What does it represent? We can't introduce a struct named `K` into the app.

File: Sources/OrdermentumSDK/Utilities/Config.swift [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    struct ConfigKey {
+        static let clientId = "CLIENT_ID"
+        static let clientSecret = "CLIENT_SECRET"
+    }
+}
+
+// MARK: - Read client config plist
+struct Config {
+    private static var main: [String: String] {
+        if let path = Bundle.main.path(forResource: K.mainConfig, ofType: "plist"), let dict = NSDictionary(contentsOfFile: path) as? [String: String] {
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Utilities/Config.swift:22]: I don't see where you've introduce the plist file into the SDK?

File: Sources/OrdermentumSDK/Models/Cart.swift (hunk 2 above)
Comment (mkmbayona) [reply Sources/OrdermentumSDK/Models/Cart.swift:19]: @bstillitano Just stumbled upon it and looks like it is not being used.

File: Sources/OrdermentumSDK/Models/Cart.swift (hunk 3 above)
Comment (mkmbayona) [reply Sources/OrdermentumSDK/Models/Cart.swift:50]: @bstillitano Yeah just stumbled upon it and looks like it is not being used.

File: Sources/OrdermentumSDK/Models/Auth.swift [hunk 9]
```diff
@@ -59,3 +59,91 @@ public struct ResendVerifyEmailRequest: Codable {
...
+    
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        id = container.tryDecode(type: String.self, forKey: .id) ?? ""
+        token = container.tryDecode(type: String.self, forKey: .token) ?? ""
+        accessToken = container.tryDecode(type: String.self, forKey: .access_token) ?? ""
+        refreshToken = container.tryDecode(type: String.self, forKey: .refresh_token) ?? ""
+        expiresIn = container.tryDecode(type: Double.self, forKey: .expires_in) ?? 0
+        tokenType = container.tryDecode(type: String.self, forKey: .token_type) ?? ""
+    }
```
[the remark this answers is not in the document]
Comment (mkmbayona) [reply Sources/OrdermentumSDK/Models/Auth.swift:122]: @bstillitano Done

File: Sources/OrdermentumSDK/Clients/AuthClient.swift [hunk 10]
```diff
@@ -21,6 +21,7 @@ public class AuthClient {
     
     //Services
     public var social: SocialService = SocialService()
+    public var auth = AuthServiceV2()
```
[the remark this answers is not in the document]
Comment (mkmbayona) [reply Sources/OrdermentumSDK/Clients/AuthClient.swift]: @bstillitano We have an existing `AuthService`. Although I think we can put this into that class as the differences are all handled within the `AuthRouter`. You think this is a good time to transfer all of the auth stuff into the `OM` class?
Happy to discuss

File: Sources/OrdermentumSDK/Utilities/Config.swift (hunk 8 above)
Comment (mkmbayona) [reply Sources/OrdermentumSDK/Utilities/Config.swift]: @bstillitano The plist will actually be on the app. Goal was to somehow be similar to GoogleService-Info.plist for Firebase.
The SDK is a local one here but I've verified this works with the swift package using the repo url + specific branch:

image

File: Sources/OrdermentumSDK/Clients/AuthClient.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (bstillitano) [reply Sources/OrdermentumSDK/Clients/AuthClient.swift]: Happy for you to migrate everything to OM.

File: Sources/OrdermentumSDK/Models/Cart.swift (hunk 2 above)
Comment (bstillitano) [reply Sources/OrdermentumSDK/Models/Cart.swift:19]: Ok. So long as it's not being used it's fine to remove.

File: Sources/OrdermentumSDK/Models/Auth.swift [hunk 11]
```diff
@@ -59,3 +59,65 @@ public struct ResendVerifyEmailRequest: Codable {
     
     public var email: String = ""
 }
+
+public struct AuthRequest: Codable {
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Models/Auth.swift:63]: All strcuts within the SDK must be initialisable without params. Please add a `public init() {}`

File: Sources/OrdermentumSDK/Models/Auth.swift [hunk 12]
```diff
@@ -59,3 +59,65 @@ public struct ResendVerifyEmailRequest: Codable {
...
+    }
+    
+    init(refreshToken: String) {
+        self.grant_type = GrantType.refreshToken.rawValue
+        
+        self.refresh_token = refreshToken
+    }
+}
+
+public struct AuthRevoke: Codable {
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Models/Auth.swift:96]: All strcuts within the SDK must be initialisable without params. Please add a public init() {}

File: Sources/OrdermentumSDK/Routers/AuthRouter.swift [hunk 13]
```diff
@@ -36,12 +38,37 @@ public enum AuthRouter: URLRequestConvertible {
...
+            return OM.instance.endpoints.auth
+        default:
+            return OM.instance.endpoints.ordermentum
         }
     }
     
     var version: String {
         switch self {
-        case .resetPassword: // Not on v1?
+        case .resetPassword,
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Routers/AuthRouter.swift:70]: As mentioned in an earlier review, please make these case by case. I understand that it's not best pracitce but in the context of the SDK, it make it easier to understand what we're hitting and also makes documentation a little nicer.

**Review by bstillitano — CHANGES_REQUESTED** [review]

File: Sources/OrdermentumSDK/Services/AuthService.swift [hunk 14]
```diff
@@ -108,4 +106,50 @@ public class AuthService {
...
+     */
+    private static func requestToken(requestBody: AuthRequest, completion: @escaping (Bool, AuthAccess?, ErrorResponse?) -> Void) {
+        let route = AuthRouter.requestToken(requestBody) as URLRequestConvertible
+        
+        Service<AuthAccess, ErrorResponse>().request(route: route) { (result, responseObject, errorObject) in
+            completion(result, responseObject, errorObject)
+        }
+    }
+    
+    // MARK: - New auth services
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Services/AuthService.swift:122]: Can we please remove this pragma.

**Review by bstillitano — CHANGES_REQUESTED** [review]

File: Sources/OrdermentumSDK/Utilities/Config.swift [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+//  File.swift
+//  
+//
+//  Created by Mark Kenneth Bayona on 9/2/21.
+//
+
+import Foundation
+
+// MARK: - Read client config plist
+struct Config {
```
Comment (bstillitano) [inline Sources/OrdermentumSDK/Utilities/Config.swift:11]: Does this need to be a struct? Looks like all properties are calculated.

**Pull request merged**